Add automatic sharding support for ttnn operation

Aswinmcw opened the pull request:
### Ticket
NA

### Problem description
Currently we can't simply pass shard config as HEIGHT_SHARDED without shard spec

### What's changed

1. Allowing sharded output tensors even when the input is not sharded, as long as a valid shard_spec can be automatically created.
2. Automatically generating an appropriate shard_spec for the output when:
- output_memory_config.is_sharded() == true
- and output_memory_config.shard_spec() is missing.
3. Ensuring the automatically created shard_spec adheres to:
- device tile alignment (32×32)
- L1 alignment constraints
- grid size limits
- correct shard orientation for WIDTH / HEIGHT / BLOCK sharding
4. Ensuring BLOCK_SHARDED properly computes shard grid shape and respects device core topology.
5. Using the input tensor’s layout for the output layout when sharded.

Perf for 1,1,1024,1024:

| Memory Config        | Device Kernel Duration (ns) | Improvement (%) |
|----------------------|-----------------------------|------------------|
| Interleaved          | 19228.00                    |            |
| L1 Width-Sharded     | 12559.00                    | 34.67%           |
| L1 Height-Sharded    | 12141.00                    | 36.84%           |
| L1 Block-Sharded     | 12452.00                    | 35.24%           |

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
Add auto sharded memory configuration support for unary operations

### Overview
Enhanced the unary device operation to allow sharded output from non-sharded input when a shard specification can be automatically created. This improves flexibility in memory management for tensor operations.

### Changes
- Updated validation logic to permit sharded output under specific conditions.
- Implemented automatic shard specification creation based on input tensor properties.
- Added a new test for the `hardmish` operation with various input shapes and memory configurations to ensure correctness.

### Testing
New tests added to validate the changes, ensuring all existing tests pass successfully.

Aswinmcw committed:
Enhance shard dimension calculations in UnaryDeviceOperation

### Overview
Refined the shard dimension calculations in the UnaryDeviceOperation to ensure that shard heights and widths do not exceed total dimensions. This improves the robustness of the output specifications computation.

### Changes
- Updated logic to enforce minimum and maximum constraints on shard dimensions.
- Simplified validation checks for sharded output conditions.
- Removed redundant calculations and comments for clarity.

### Testing
No new tests added; existing tests cover the modified functionality, ensuring correctness.

Aswinmcw committed:
Enhance validation for sharding configurations in UnaryDeviceOperation

### Overview
Improved the validation logic for sharding configurations in the UnaryDeviceOperation to ensure that the total dimensions meet the required constraints for both width and height sharding. This enhances the robustness of the output specifications computation.

### Changes
- Added TT_FATAL assertions to validate that total_width and total_height meet minimum requirements for WIDTH_SHARDED and HEIGHT_SHARDED configurations, respectively.
- Simplified the logic for limiting the number of cores to the available grid size.

### Testing
No new tests added; existing tests cover the modified functionality, ensuring correctness.

dchenTT commented:
It is nice. But I am curious about the user cases of calculating shard spec in host vs in device. If done in host, it can apply to any op. Input is tensor shape, sharding strategy, assuming use all cores, output is shard size and shard spec.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/unary_auto_shard

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/unary_auto_shard

Aswinmcw committed:
Add automatic shard specification computation for MemoryConfig

This update introduces the `compute_auto_shard_spec` function, which automatically computes and fills in the shard specification for a MemoryConfig if it is sharded but missing a shard_spec. The function reuses the input tensor's shard_spec if available, or computes a new one based on the input tensor's shape, memory layout, and device grid. This change enhances the flexibility and usability of tensor operations by ensuring proper shard configurations are applied when needed.

dchenTT commented on ttnn/core/tensor/tensor_utils.cpp:
```diff
@@ -108,6 +111,224 @@ ShardDivisionSpec compute_shard_division_spec(const Shape2D& shape, const Shape2
+
+    switch (memory_layout) {
+        case TensorMemoryLayout::WIDTH_SHARDED: {
+            // For width sharding, maximize parallelism while ensuring alignment
```
for all sharding strategy, do you consider uneven shards? Since for existing implementation, uneven shard has a big performance penalty (using direct L1 sharding vs tensor accessor), even shard has better performance even if number of cores are smaller.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/unary_auto_shard

Aswinmcw commented:
@dchenTT i have one doubt, we are passing the mem config while calling the op, so the input is still interleaved, do we need to convert the tensor to sharded using interleaved_to_sharded then perform the operation?
